core/src/main/java/hivemall/mf/CofactorModel.java

```diff
@@ -0,0 +1,629 @@
+                predicted = contextFactors.dotProduct(unratedFactors);
+                err = (0.d - predicted);
+                loss += c0 * err * err;
+            }
+            // only rated items / users
+            for (Feature f : sample.features) {
+                if (!isTrainable(f.getFeature(), featureWeights)) {
+                    continue;
+                }
+                ratedFactors = getFactorVector(f.getFeature(), featureWeights);
+                predicted = contextFactors.dotProduct(ratedFactors);
+                y = f.getValue();
+                err = y - predicted;
+                loss += (c1 - c0) * err * err;
+            }
+        }
+        return loss;
+    }
+
+    protected static double sumL2Loss(Map<String, RealVector> weights, float lambda) {
+        double loss = 0.d;
+        for (RealVector v : weights.values()) {
+            loss += L2Distance(v);
+        }
+        return lambda * loss;
+    }
+
+    protected static double L2Distance(RealVector v) {
+        double result = 0.d;
+        for (int i = 0; i < v.getDimension(); i++) {
+            double val = v.getEntry(i);
+            result +=  val * val;
+        }
+        return Math.sqrt(result);
+    }
+
+    /**
+     * Add v to u in-place without creating a new RealVector instance.
+     * @param u vector to which v will be added
+     * @param v vector containing new values to be added to u
+     * @param scalar value to multiply each entry in v before adding to u
+     */
+    private static void addInPlace(RealVector u, RealVector v, double scalar) {
+        assert u.getDimension() == v.getDimension();
+        for (int i = 0; i < u.getDimension(); i++) {
+            double newVal = u.getEntry(i) + scalar * v.getEntry(i);
+            u.setEntry(i, newVal);
+        }
+    }
+
+    private static boolean isTrainable(String name, Map<String, RealVector> weights) {
+        return weights.containsKey(name);
+    }
+
+    @Nonnull
+    private static Random[] newRandoms(@Nonnull final int size, final long seed) {
+        final Random[] rand = new Random[size];
+        for (int i = 0, len = rand.length; i < len; i++) {
+            rand[i] = new Random(seed + i);
+        }
+        return rand;
+    }
+
+    protected static void uniformFill(final RealVector a, final Random rand, final float maxInitValue) {
+        for (int i = 0, len = a.getDimension(); i < len; i++) {
+            double v = rand.nextDouble() * maxInitValue / len;
+            a.setEntry(i, v);
+        }
+    }
+
+    protected static void gaussianFill(final RealVector a, final Random[] rand, final double stddev) {
+        for (int i = 0, len = a.getDimension(); i < len; i++) {
+            double v = MathUtils.gaussian(0.d, stddev, rand[i]);
+            a.setEntry(i, v);
+        }
+    }
+
+    private static void checkHyperparameterC(final float c) {
+        assert c >= 0.f && c <= 1.f;
```

Comment:
Better to throw exceptions instead of assertion (could be off in production).

`Precondition.checkArgument(c >= 0.f && c <= 1.f)`